### ISSUE TITLE
'Vector{nnn}.Create' methods have non ideal codegen for `CreateScalar`

### ISSUE THREAD

**Issue by john-h-k** [dotnet/coreclr#27015]
Current codegen for `Vector128.CreateScalar(float|double)` and `Vector256.CreateScalar(float|double)` is sub par.
Ideal codegen would be

```x86asm
vxorps xmm0, xmm0, xmm0
vmovss xmm0, [src]```

however, the [currently generated code](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAJI8ovLrl5hcAbhr0mrTj36CRGMRKm4WADQAcSOdXkBmBt1zYAZjCakGAYQY0A3jQaem55kgYA1GDAMaHJSfQAeawAbCGwMAD4nWHiYAGUwbGjsKDD9AAoYuIwAKgYABwBKF2ovBndauq9iAHYAoJDc8JZHFIx0zOyofJKq4zqAXw8vac9iH3I/QODoUgBWJCjY+KTemFSMrJz1pELt0orq2frrutb2lagTnr6Bo+HRyvGvKeoJoA==) for `Vector128.CreateScalar(float)` is

```x86asm
vmovss xmm0, dword [rdx]
vxorps xmm1, xmm1, xmm1
vmovss xmm0, xmm1, xmm0```

(exact same for ideal and generated with `double`, just `sd/pd` instead

which is less efficient.

This also applies to `Vector256.CreateScalar(float|double)`.

**Comment by EgorBo** [comment]
Why are you passing float by ref/pointer? If you pass it by value it will be passed in xmm0 (in your case)
Also, there is `Vector128.CreateScalarUnsafe`

**Comment by john-h-k** [comment]
@EgorBo 

> Also, there is Vector128.CreateScalarUnsafe

Yes, but I want the last 3 elements to be zero'd, so that method is not suitable.

>  If you pass it by value it will be passed in xmm0 (in your case)

That is rather the issue. `Vector128,Create` does not merge the dereference of the pointer and the creation, as would be ideal (or having `Vector128.Create(float*)`)

**Comment by BruceForstall** [comment]
@tannergooding @CarolEidt  @echesakovMSFT

**Comment by tannergooding** [comment]
This is basically a dupe of https://github.com/dotnet/coreclr/issues/17225 (although the older one is from when we were calling the methods `SetVector128`).

The fix should probably be basically the same as https://github.com/dotnet/coreclr/pull/14393

It would, however, require us to either move the non-constant fallback into codegen (it is in managed currently) or to determine how we are addressing https://github.com/dotnet/coreclr/issues/19888 more generally...

**Comment by tannergooding** [comment]
@CarolEidt, do you have an opinion on a good way to address the constant folding problem initially?

That is, we have some hwintrinsics that require (or greatly benefit) from their inputs being constant. Those can't easily be handled entirely in importation, and so we need to try to handle them in lowering as that catches the most cases...

We could keep the call in the importer and then try to lower it to the relevant intrinsic chain again if the required args are constant by the time we get to lowering. I think we could also also always expand in importation and then lower back to the relevant call if they don't get transformed appropriately (although I think this is more difficult to handle and I don't believe we have any code that does this today)...
